pkg/pipeline/write/write_tcp.go

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2021 IBM, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package write
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+
+	"github.com/netobserv/flowlogs-pipeline/pkg/config"
+	"github.com/sirupsen/logrus"
+)
+
+type writeTCP struct {
+	address string
+	conn    net.Conn
+}
+
+// Write writes a flow to tcp connection
+func (t *writeTCP) Write(v config.GenericMap) {
+	logrus.Tracef("entering writeTCP Write")
+	b, _ := json.Marshal(v)
+	// append new line between each record to split on client side
+	b = append(b, []byte("\n")...)
+	_, err := t.conn.Write(b)
+	if err != nil {
+		log.WithError(err).Warn("can't write tcp")
+	}
+}
+
+// NewWriteTCP create a new write
+func NewWriteTCP(params config.StageParam) (Writer, error) {
+	logrus.Debugf("entering NewWriteTCP")
+	writeTCP := &writeTCP{}
+	if params.Write != nil && params.Write.TCP != nil && params.Write.TCP.Port != "" {
+		writeTCP.address = ":" + params.Write.TCP.Port
+	} else {
+		return nil, fmt.Errorf("Write.TCP.Port must be specified")
+	}
+
+	logrus.Debugf("NewWriteTCP Listen %s", writeTCP.address)
+	l, err := net.Listen("tcp", writeTCP.address)
+	if err != nil {
+		return nil, err
+	}
+	defer l.Close()
+	clientConn, err := l.Accept()
```

Comment:
It looks like you are expecting some other entity to initiate a TCP connection to FLP to which FLP will write its flow logs.
The code will block here until a single connection is completed. The rest of FLP will wait on this connection. Is this the desired behavior? 

What is the use case you have in mind? 

What if I have multiple targets to which I want to send my data via TCP? Can I set up multiple writeTCP stages? How will it work? or do we want to put the listen() operation into a separate thread and accumulate the connections that were attempted and send flow logs to all current connections?

Reply:
This is a temporary PR used in netobserv CLI: https://github.com/netobserv/network-observability-cli/pull/2#discussion_r1497191885

The CLI deploys eBPF agents using `direct-flp` mode which embedd flowlogs-pipeline into the same pods.
Currently the connection between these pods and cli is made [using TCP](https://github.com/netobserv/network-observability-cli/blob/main/cmd/flow-capture.go#L68) after calling [oc port-forward](https://github.com/netobserv/network-observability-cli/blob/main/res/functions.sh#L37-L57) commands.
Using gRPC instead is still in discussion.

If we keep that TCP implementation, we can indeed improve and makes the wait optionnal + allow multiple targets :+1: